source/browseMode.py

```diff
@@ -477,10 +488,51 @@ def _activatePosition(self,obj=None):
 			self._activateNVDAObject(obj)
 
 	def script_activatePosition(self,gesture):
-		self._activatePosition()
+		self.maybeSyncFocus(postFocusFunc=self._activatePosition)

```

Comment:
Can you explain what would happen if activatePosition was executed straight away possibly before setfocus completes? I.e. why the need for the asynchronis stuff?

Reply:
Good question - I expected you to ask it :-)

The problem I was facing was that when you click on a checkbox on a web page, its updated status won't be announced. This also applies to radio buttons and potentially other form controls.

Here are what my investigation has shown:
Typically in NVDA speech.speakObjectProperties is called twice:
* First, when a checkbox is focused, e.g. browse mode cursor is on it. It is called with reason=controlTypes.REASON_ONLYCACHE, which doesn't actuallyspeak, but only remembers its state. 
* Second, when script_activatePosition() is called,, e.g. user pressed enter on a checkbox, speakObjectProperties is called with reason="change", which compares all the properties of NVDAObject and only speaks whatever is changed.

With FastBrowseMode, Focus event and activate event are issued very closely together. Since both of them are processed asynchronously, via event dispatcher, focus processing (browsermode.event_gainFocus()) is actually called after activatePosition(), because it is called asynchronously and naturally there is some delay. As a result, it sees already updated state of checkbox. As a result, the next call to speak does nothing, because from its point of view nothing has changed.

The way to resolve this problem is to make sure that focus event and all the postprocessing happens stricly before activatePosition call is issued. So we have to wait until async event onFocusGain  comes in and is fully processed.